assets/arc-0052/bip32-ed25519.ts

```diff
@@ -0,0 +1,181 @@
+ * and
+ * C) Private Child Key Derivation
+ * 
+ * @param extendedKey - extended key (kL, kR, c) where kL is the left 32 bytes of the root key the scalar (pvtKey). kR is the right 32 bytes of the root key, and c is the chain code. Total 96 bytes
+ * @param index - index of the child key
+ * @returns - (kL, kR, c) where kL is the left 32 bytes of the child key (the new scalar), kR is the right 32 bytes of the child key, and c is the chain code. Total 96 bytes
+ */
+export function deriveChildNodePrivate(
+  extendedKey: Uint8Array,
+  index: number
+): Uint8Array {
+  const kl: Buffer = Buffer.from(extendedKey.subarray(0, 32));
+  const kr: Buffer = Buffer.from(extendedKey.subarray(32, 64));
+  const cc: Uint8Array = extendedKey.subarray(64, 96);
+
+  const { z, childChainCode } = index < 0x80000000 ? derivedNonHardened(kl, cc, index) : deriveHardened(kl, kr, cc, index);
+
+  const chainCode = childChainCode.subarray(32, 64);
+  const zl = z.subarray(0, 32);
+  const zr = z.subarray(32, 64);
+
+  // left = kl + 8 * trunc28(zl)
+  // right = zr + kr
+
+  const left = new BN(kl, 16, "le").add(new BN(zl.subarray(0, 28), 16, "le").mul(new BN(8))).toArrayLike(Buffer, "le", 32);
+  let right = new BN(kr, 16, "le").add(new BN(zr, 16, "le")).toArrayLike(Buffer, "le").slice(0, 32);
+
+  // just padding
+  if (right.length !== 32) {
+    right = Buffer.from(right.toString("hex") + "00", "hex");
+  }
+
+  return Buffer.concat([left, right, chainCode]);
+}
+
+/**
+ *  * @see section V. BIP32-Ed25519: Specification;
+ * 
+ * subsections:
+ * 
+ * D) Public Child key
+ * 
+ * @param extendedKey - extend public key (p, c) where p is the public key and c is the chain code. Total 64 bytes
+ * @param index - unharden index (i < 2^31) of the child key
+ * @returns - 64 bytes, being the 32 bytes of the child key (the new public key) followed by the 32 bytes of the chain code
+ */
+export function deriveChildNodePublic(extendedKey: Uint8Array, index: number): Uint8Array {
+    if (index > 0x80000000) throw new Error('can not derive public key with harden')
+
+    const pk: Buffer = Buffer.from(extendedKey.subarray(0, 32))
+    const cc: Buffer = Buffer.from(extendedKey.subarray(32, 64))
+
+    const data: Buffer = Buffer.allocUnsafe(1 + 32 + 4);
+    data.writeUInt32LE(index, 1 + 32);
+
+    pk.copy(data, 1);
+    data[0] = 0x02;
+        
+    const z: Buffer = createHmac("sha512", cc).update(data).digest();
+    data[0] = 0x03;
+    
+    const i: Buffer = createHmac("sha512", cc).update(data).digest();
+
+    // Section V. BIP32-Ed25519: Specification; subsection D) Public Child Key Derivation
+    const chainCode: Buffer = i.subarray(32, 64);
+    const zl: Buffer = z.subarray(0, 32);
+
+    // left = 8 * 28bytesOf(zl)
+    const left = new BN(zl.subarray(0, 28), 16, 'le').mul(new BN(8)).toArrayLike(Buffer, 'le', 32);
+
+    const p: Uint8Array = crypto_scalarmult_ed25519_base_noclamp(left);
+    return Buffer.concat([crypto_core_ed25519_add(p, pk), chainCode]);
```

Comment:
In the specification it says `If Ai is the identity point (0,1), discard the child`. Where do we check for this?